File: tests/cpu/python/test_print.py
"""Print CPU binding tests — aligned with C++ test_print.cpp.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python
    LD_LIBRARY_PATH=build/backends/cpu
    python -m pytest tests/cpu/python/test_print.py -v
"""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)


class TestPrintCPU:
    """Print / repr / str operations — CPU backend."""

    def test_repr_1d(self):
        a = ins.zeros([5], ins.float32)
        r = repr(a)
        assert "insight" in r.lower() or "array" in r.lower() or "float32" in r

    def test_repr_2d(self):
        a = ins.ones([3, 4], ins.float64)
        r = repr(a)
        assert isinstance(r, str) and len(r) > 0

    def test_repr_int(self):
        a = ins.zeros([3], ins.int32)
        r = repr(a)
        assert isinstance(r, str) and len(r) > 0

    def test_str_1d(self):
        a = ins.from_numpy(np.array([1, 2, 3], dtype=np.float32))
        s = str(a)
        assert isinstance(s, str) and len(s) > 0

    def test_str_2d(self):
        a = ins.from_numpy(np.array([[1, 2], [3, 4]], dtype=np.float64))
        s = str(a)
        assert isinstance(s, str) and len(s) > 0

    def test_print_does_not_crash_scalar(self):
        a = ins.zeros([], ins.float32)
        # Should not raise
        _ = repr(a)

    def test_print_does_not_crash_large(self):
        a = ins.ones([100, 100], ins.float32)
        _ = repr(a)

    def test_print_empty(self):
        a = ins.zeros([0], ins.float32)
        _ = repr(a)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cpu/lua/test_print.lua
-- Print CPU binding tests — aligned with C++ test_print.cpp
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;"
--   LD_LIBRARY_PATH=build/backends/cpu ~/.luarocks/bin/busted tests/cpu/lua/test_print.lua

describe("Print CPU Tests", function()
  local ins
  setup(function()
    ins = require("_insight")
    ins.set_device(ins.CPUPlace())
  end)

  it("tostring_1d", function()
    local a = ins.zeros({ 5 }, ins.float32)
    local s = tostring(a)
    assert.is_not_nil(s)
    assert.is_true(#s > 0)
  end)

  it("tostring_2d", function()
    local a = ins.ones({ 3, 4 }, ins.float64)
    local s = tostring(a)
    assert.is_not_nil(s)
    assert.is_true(#s > 0)
  end)

  it("tostring_int", function()
    local a = ins.zeros({ 3 }, ins.int32)
    local s = tostring(a)
    assert.is_not_nil(s)
    assert.is_true(#s > 0)
  end)

  it("tostring_from_table", function()
    local a = ins.from_table({ 1.0, 2.0, 3.0 })
    local s = tostring(a)
    assert.is_not_nil(s)
    assert.is_true(#s > 0)
  end)

  it("tostring_2d_table", function()
    local a = ins.from_table({ { 1, 2 }, { 3, 4 } })
    local s = tostring(a)
    assert.is_not_nil(s)
    assert.is_true(#s > 0)
  end)

  it("repr_does_not_crash_large", function()
    local a = ins.ones({ 100, 100 }, ins.float32)
    local s = tostring(a)
    assert.is_not_nil(s)
  end)

  it("repr_empty", function()
    local a = ins.zeros({ 0 }, ins.float32)
    local s = tostring(a)
    assert.is_not_nil(s)
  end)

  it("repr_after_cast", function()
    local a = ins.ones({ 3 }, ins.float32)
    local b = ins.cast(a, ins.int32)
    local s = tostring(b)
    assert.is_not_nil(s)
  end)
end)
